source/MRViewer/MRSurfaceContoursWidget.cpp

```diff
@@ -24,93 +24,55 @@ void updateBaseColor( std::shared_ptr<SurfacePointWidget> point, const Color& co
-
-    auto& contour = widget_.pickedPoints_[obj_];
-    if ( actionType == Type::Undo )
-    {
-        widget_.surfacePointWidgetCache_.erase( contour.back()->getPickSphere().get() );
-        contour.pop_back();
-
-        widget_.highlightLastPoint( obj_ );
-        widget_.activeIndex_ = int( contour.size() - 1 );
-        widget_.activeObject_ = obj_;
-
-        widget_.onPointRemove_( obj_ );
-    }
-    else
-    {
-        contour.push_back( widget_.createPickWidget_( obj_, point_ ) );
-
-        widget_.activeIndex_ = int( contour.size() - 1 );
-        widget_.activeObject_ = obj_;
-        widget_.highlightLastPoint( obj_ );
-
-        widget_.onPointAdd_( obj_ );
-
-    }
-
-}
-size_t SurfaceContoursWidget::AddPointActionPickerPoint::heapBytes() const
-{
-    return 0; //this undo action will be deleted in widget disable
-}
+    /// removes point by index, and returns undo action for its addition at the same place
+    static std::shared_ptr<AddRemovePointHistoryAction> remove( SurfaceContoursWidget& widget, const std::shared_ptr<MR::VisualObject>& obj, int index );
 
+    virtual std::string name() const override { return name_ + widget_.params.historyNameSuffix; }
+    virtual void action( Type actionType ) override;
+    [[nodiscard]] virtual size_t heapBytes() const override { return 0; } //this undo action will be deleted in widget disable
 
-class SurfaceContoursWidget::RemovePointActionPickerPoint : public SurfaceContoursWidget::WidgetHistoryAction
-{
-public:
-    RemovePointActionPickerPoint( SurfaceContoursWidget& widget, const std::shared_ptr<MR::VisualObject>& obj, const PickedPoint& point, int index ) :
+private:
+    AddRemovePointHistoryAction( std::string name, SurfaceContoursWidget& widget, const std::shared_ptr<MR::VisualObject>& obj, const PickedPoint& point, int index, bool insertOnAction ) :
+        name_{ std::move( name ) },
         widget_{ widget },
         obj_{ obj },
         point_{ point },
-        index_{ index }
+        index_{ index },
+        insertOnAction_{ insertOnAction }
     {};
 
-    virtual std::string name() const override;
-    virtual void action( Type actionType ) override;
-    [[nodiscard]] virtual size_t heapBytes() const override;
-private:
+    std::string name_;
     SurfaceContoursWidget& widget_;
     const std::shared_ptr<MR::VisualObject> obj_;
     PickedPoint point_;
-    int index_;
+    int index_ = 0; // -1 here means insert after last one
+    bool insertOnAction_ = false;
 };
 
-std::string SurfaceContoursWidget::RemovePointActionPickerPoint::name() const
+std::shared_ptr<SurfaceContoursWidget::AddRemovePointHistoryAction> SurfaceContoursWidget::AddRemovePointHistoryAction::append(
+    SurfaceContoursWidget& widget, const std::shared_ptr<MR::VisualObject>& obj, const PickedPoint& point )
+{
+    std::shared_ptr<AddRemovePointHistoryAction> res( new AddRemovePointHistoryAction( "Append Point", widget, obj, point, -1, true ) );
+    res->action( Type::Redo );
```

Comment:
Maybe better to add `append` and `remove` methods and call them depending on the action's state.

Reply:
ok